File: src/main/java/net/wanhack/service/region/generators/RoomFirstRegionGenerator.java
/*
 *  Copyright 2005 The Wanhack Team
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package net.wanhack.service.region.generators;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import net.wanhack.model.common.Direction;
import net.wanhack.model.region.Cell;
import net.wanhack.model.region.CellSet;
import net.wanhack.model.region.CellType;
import net.wanhack.model.region.Door;
import net.wanhack.model.region.Region;
import net.wanhack.model.region.World;
import net.wanhack.utils.Probability;
import net.wanhack.utils.RandomUtils;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;


/**
 * A simple region generator that first generates some rooms on the
 * region and then proceeds to make corridors between the rooms.
 * <p>
 * The algorithm starts by creating minRooms-maxRooms rooms of random
 * size on random places of the region. The algorithm tries to make the
 * rooms not overlap by trying to randomize the room <code>overlapTries</code>
 * times if it overlaps with existing rooms.
 * <p>
 * When the rooms have randomized and placed on map, the next step is to
 * make corridors between them. Two rooms are chosen at random, and a corridor
 * is made between them. If the corridor bumps into another room or the target,
 * then the corridor will end there. Then the next two rooms are chosen, and
 * a new corridor is created etc. This process will continue until all rooms
 * are reachable from all others.
 */
public class RoomFirstRegionGenerator implements RegionGenerator {

    private Region region;
    private int width;
    private int height;
    
    private int minRooms = 4;
    private int maxRooms = 10;
    private int roomMinWidth = 6;
    private int roomMaxWidth = 18;
    private int roomMinHeight = 5;
    private int roomMaxHeight = 9;
    private final Probability connectConnectedProbability = new Probability(50);
    private final Probability doorProbability = new Probability(30);
    private final Probability hiddenDoorProbability = new Probability(15);
    private final int overlapTries = 20;
    private final Random random = new Random();
    
    private final Log log = LogFactory.getLog(getClass());
    
    public Region generate(World world, String name, int level, String up, String down) {
        initRegionParameters(level);
        
        this.region = new Region(world, name, level, width, height);
        
        List<Room> rooms = createRooms();
        createCorridors(rooms);
        createDoors();
        addStairsUpAndDown(rooms, up, down);
        return region;
    }
    
    private void initRegionParameters(int level) {
        if (level < 5) {
            width = 80;
            height = 25;
            minRooms = 4;
            maxRooms = 10;
            roomMinWidth = 6;
            roomMaxWidth = 18;
            roomMinHeight = 5;
            roomMaxHeight = 9;
            
        } else if (level < 10) {
            width = 120;
            height = 30;
            minRooms = 6;
            maxRooms = 12;
            roomMinWidth = 6;
            roomMaxWidth = 18;
            roomMinHeight = 5;
            roomMaxHeight = 9;
            
        } else if (level < 20) {
            width = 160;
            height = 40;
            minRooms = 8;
            maxRooms = 16;
            roomMinWidth = 6;
            roomMaxWidth = 18;
            roomMinHeight = 5;
            roomMaxHeight = 9;

        } else {
            width = 200;
            height = 50;
            minRooms = 10;
            maxRooms = 25;
            roomMinWidth = 6;
            roomMaxWidth = 18;
            roomMinHeight = 5;
            roomMaxHeight = 9;
        }
    }

    private List<Room> createRooms() {
        int roomCount = minRooms + random.nextInt(maxRooms - minRooms + 1);
        
        List<Room> rooms = new ArrayList<Room>(roomCount);
        for (int i = 0; i < roomCount; i++) {
            rooms.add(createRoom());
        }
        
        return rooms;
    }

    private Room createRoom() {
        Room room = randomRoom();
        int tries = overlapTries;
        
        while (overlapsExisting(room) && tries-- > 0) {
            room = randomRoom();
        }
        
        room.addToRegion();
        return room;
    }
    
    private void createDoors() {
        for (Cell cell : region) {
            if (isDoorCandidate(cell) && doorProbability.check()) {
                boolean hidden = hiddenDoorProbability.check();
                cell.setState(new Door(hidden));
            }
        }
    }

    private boolean isDoorCandidate(Cell cell) {
        if (cell.getType() != CellType.HALLWAY_FLOOR) {
            return false;
        }
        
        Cell roomNeighbour = null;
        Cell hallwayNeighbour = null;
        int walls = 0;
        for (Cell neighbour : cell.getAdjacentCellsInMainDirections()) {
            if (neighbour.getType() == CellType.ROOM_FLOOR) {
                roomNeighbour = neighbour;
            } else if (neighbour.getType() == CellType.HALLWAY_FLOOR) {
                hallwayNeighbour = neighbour;
            } else if (neighbour.getType() == CellType.ROOM_WALL) {
                walls++;
            }
        }
        
        if (roomNeighbour != null && hallwayNeighbour != null && walls == 2) {
            
            Direction room = cell.getDirection(roomNeighbour);
            Direction hall = cell.getDirection(hallwayNeighbour);
            if (room.isOpposite(hall)) {
                return true;
            }
        }
           
        return false;
    }

    private void createCorridors(List<Room> rooms) {
        for (int count = 0; !allConnected(rooms); count++) {
            Room room1 = RandomUtils.randomItem(rooms);
            Room room2 = random(rooms, room1);
            
            if (!connected(room1, room2) || connectConnected()) {
                connect(room1, room2);
            }
            
            if (count > 1000) {
                log.warn("Count exceeded, bailing out.");
                break;
            }
        }
    }

    private boolean connectConnected() {
        return connectConnectedProbability.check();
    }

    private void connect(Room start, Room goal) {
        Cell startCell = start.getRandomCell(random);
        Cell goalCell = goal.getRandomCell(random);
        
        Cell previous = null;
        for (Cell cell : createPath(startCell, goalCell)) {
            if (cell.getType() != CellType.ROOM_FLOOR) {
                cell.setType(CellType.HALLWAY_FLOOR);
            }
            
            if (!start.contains(cell)) {
                for (Cell adjacent : cell.getAdjacentCellsInMainDirections()) {
                    if (adjacent != previous && adjacent.isPassable()) {
                        return;
                    }
                }
            }
            
            previous = cell;
        }
    }

    private List<Cell> createPath(Cell start, Cell goal) {
        return new CorridorPathSearcher(region).findShortestPath(start, goal);
    }

    private <T> T random(List<T> items, T invalid) {
        T result = invalid;
        while (result == invalid) {
            result = items.get(random.nextInt(items.size()));
        }
        return result;
    }
    
    private boolean allConnected(List<Room> rooms) {
        Room start = rooms.get(0);
        for (Room room : rooms) {
            if (!connected(start, room)) {
                return false;
            }
        }
        return true;
    }
    
    private boolean connected(Room room1, Room room2) {
        return room1.getMiddleCell().isReachable(room2.getMiddleCell());
    }

    private boolean overlapsExisting(Room room) {
        for (int yy = 0; yy < room.h; yy++) {
            for (int xx = 0; xx < room.w; xx++) {
                Cell cell = region.getCell(room.x + xx, room.y + yy);
                if (cell.getType() != CellType.WALL) {
                    return true;
                }
            }
        }
        return false;
    }

    public void addStairsUpAndDown(List<Room> rooms, String upRegion, String downRegion) {
        Room stairsUpRoom = RandomUtils.randomItem(rooms);
        // TODO: copy paste from MazeRegionGenerator
        
        CellSet empty = region.getRoomFloorCells();
        if (empty.size() < 2) {
            throw new IllegalStateException("not enough empty cells to place stairs");
        }
        
        Cell stairsUp = stairsUpRoom.getRandomCell(random);
        
        stairsUp.setType(CellType.STAIRS_UP);
        if (upRegion != null) {
            region.addPortal(stairsUp.x, stairsUp.y, upRegion, "from down", true);
        }
        region.addStartPoint("from up", stairsUp.x, stairsUp.y);

        while (true) {
            Room stairsDownRoom = random(rooms, stairsUpRoom);
            Cell stairsDown = stairsDownRoom.getRandomCell(random);
            if (stairsDown != stairsUp && region.findPath(stairsUp, stairsDown) != null) {
                stairsDown.setType(CellType.STAIRS_DOWN);
                region.addPortal(stairsDown.x, stairsDown.y, downRegion, "from up", false);
                region.addStartPoint("from down", stairsDown.x, stairsDown.y);
                return;
            }
        }
    }

    private Room randomRoom() {
        // TODO: copy-paste from MazeRegionGenerator
        int w = roomMinWidth + random.nextInt(1 + roomMaxWidth - roomMinWidth);
        int h = roomMinHeight + random.nextInt(1 + roomMaxHeight - roomMinHeight);
        int x = 1 + random.nextInt(width - w - 2);
        int y = 1 + random.nextInt(height - h - 2);
        
        return new Room(region, x, y, w, h);
    }
    
    private static class Room {
        private final Region region;
        private final int x;
        private final int y;
        private final int w;
        private final int h;
        
        public Room(Region region, int x, int y, int w, int h) {
            this.region = region;
            this.x = x;
            this.y = y;
            this.w = w;
            this.h = h;
        }
        
        public boolean contains(Cell cell) {
            return cell.x >= x 
                && cell.x < x + w
                && cell.y >= y 
                && cell.y < y + h;
        }
        
        public Cell getRandomCell(Random random) {
            int xx = x + 1 + random.nextInt(w - 2);
            int yy = y + 1 + random.nextInt(h - 2);
            
            return region.getCell(xx, yy);
        }

        public Cell getMiddleCell() {
            int xx = x + w / 2;
            int yy = y + h / 2;
            
            return region.getCell(xx, yy);
        }

        private void addToRegion() {
            for (int yy = 1; yy < h - 1; yy++) {
                for (int xx = 1; xx < w - 1; xx++) {
                    region.getCell(x + xx, y + yy).setType(CellType.ROOM_FLOOR);
                }
            }
            
            for (int xx = 0; xx < w; xx++) {
                region.getCell(x + xx, y).setType(CellType.ROOM_WALL);
                region.getCell(x + xx, y + h - 1).setType(CellType.ROOM_WALL);
            }
            
            for (int yy = 0; yy < h; yy++) {
                region.getCell(x, y + yy).setType(CellType.ROOM_WALL);
                region.getCell(x + w - 1, y + yy).setType(CellType.ROOM_WALL);
            }
        }
    }
}

File: src/main/kotlin/net/wanhack/model/region/generators/RoomFirstRegionGenerator.kt
/*
 * Copyright 2013 The Wanhack Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.wanhack.service.region.generators

import java.util.Random
import net.wanhack.model.region.Cell
import net.wanhack.model.region.CellType
import net.wanhack.model.region.Door
import net.wanhack.model.region.Region
import net.wanhack.model.region.World
import net.wanhack.utils.Probability
import org.apache.commons.logging.LogFactory
import net.wanhack.service.region.generators.RoomFirstRegionGenerator.RegionParameters
import net.wanhack.utils.collections.randomElement

class RoomFirstRegionGenerator(val world: World, val name: String, val level: Int, val rp: RegionParameters, val up: String?, val down: String?) {
    private val region = Region(world, name, level, rp.width, rp.height)
    private val connectConnectedProbability = Probability(50)
    private val doorProbability = Probability(30)
    private val hiddenDoorProbability = Probability(15)
    private val overlapTries: Int = 20
    private val random = Random()
    private val log = LogFactory.getLog(javaClass)

    fun generate(): Region {
        val rooms = createRooms()
        createCorridors(rooms)
        createDoors()
        addStairsUpAndDown(rooms)
        return region
    }

    private fun createRooms(): List<Room> {
        val roomCount = rp.minRooms + random.nextInt(rp.maxRooms - rp.minRooms + 1)
        val rooms = listBuilder<Room>()

        roomCount.times {
            rooms.add(createRoom())
        }

        return rooms.build()
    }

    private fun createRoom(): Room {
        var room = randomRoom()
        var tries = overlapTries
        while (overlapsExisting(room) && tries-- > 0) {
            room = randomRoom()
        }
        room.addToRegion()
        return room
    }

    private fun createDoors() {
        for (cell in region) {
            if (isDoorCandidate(cell) && doorProbability.check()) {
                val hidden = hiddenDoorProbability.check()
                cell.state = Door(hidden)
            }
        }
    }

    private fun isDoorCandidate(cell: Cell): Boolean {
        if (cell.cellType != CellType.HALLWAY_FLOOR)
            return false

        var roomNeighbour: Cell? = null
        var hallwayNeighbour: Cell? = null
        var walls: Int = 0
        for (neighbour in cell.adjacentCellsInMainDirections) {
            if (neighbour.cellType == CellType.ROOM_FLOOR)
                roomNeighbour = neighbour
            else if (neighbour.cellType == CellType.HALLWAY_FLOOR)
                hallwayNeighbour = neighbour
            else if (neighbour.cellType  == CellType.ROOM_WALL)
                walls++
        }

        if (roomNeighbour != null && hallwayNeighbour != null && walls == 2) {
            val room = cell.getDirection(roomNeighbour!!)!!
            val hall = cell.getDirection(hallwayNeighbour!!)!!
            if (room.isOpposite(hall))
                return true
        }

        return false
    }

    private fun createCorridors(rooms: List<Room>) {
        var count = 0
        while (!allConnected(rooms)) {
            val room1 = rooms.randomElement()
            val room2 = random(rooms, room1)
            if (!connected(room1, room2) || connectConnected())
                connect(room1, room2)

            if (count > 1000) {
                log.warn("Count exceeded, bailing out.")
                break
            }
            count++
        }
    }

    private fun connectConnected() =
        connectConnectedProbability.check()

    private fun connect(start: Room, goal: Room) {
        val startCell = start.getRandomCell(random)
        val goalCell = goal.getRandomCell(random)
        var previous: Cell? = null
        for (cell in createPath(startCell, goalCell)!!) {
            if (cell.cellType != CellType.ROOM_FLOOR)
                cell.setType(CellType.HALLWAY_FLOOR)

            if (cell !in start)
                for (adjacent in cell.adjacentCellsInMainDirections)
                    if (adjacent != previous && adjacent.isPassable())
                        return

            previous = cell
        }
    }
    private fun createPath(start: Cell, goal: Cell): List<Cell>? =
        CorridorPathSearcher(region).findShortestPath(start, goal)

    private fun random<T>(items: List<T>, invalid: T): T {
        var result = invalid
        while (result == invalid)
            result = items.randomElement()

        return result
    }

    private fun allConnected(rooms: List<Room>) =
        rooms.all { connected(rooms[0], it) }

    private fun connected(room1: Room, room2: Room) =
        room1.getMiddleCell().isReachable(room2.getMiddleCell())

    private fun overlapsExisting(room: Room): Boolean {
        for (yy in 0..room.h - 1) {
            for (xx in 0..room.w - 1) {
                val cell = region.getCell(room.x + xx, room.y + yy)
                if (cell.cellType != CellType.WALL)
                    return true
            }
        }
        return false
    }

    fun addStairsUpAndDown(rooms: List<Room>) {
        val stairsUpRoom = rooms.randomElement()
        val empty = region.getRoomFloorCells()
        if (empty.size < 2)
            throw IllegalStateException("not enough empty cells to place stairs")

        val stairsUp = stairsUpRoom.getRandomCell(random)
        stairsUp.setType(CellType.STAIRS_UP)
        if (up != null)
            region.addPortal(stairsUp.x, stairsUp.y, up, "from down", true)

        region.addStartPoint("from up", stairsUp.x, stairsUp.y)

        if (down != null) {
            while (true)
            {
                val stairsDownRoom = random(rooms, stairsUpRoom)
                val stairsDown = stairsDownRoom.getRandomCell(random)
                if (stairsDown != stairsUp && region.findPath(stairsUp, stairsDown) != null) {
                    stairsDown.setType(CellType.STAIRS_DOWN)
                    region.addPortal(stairsDown.x, stairsDown.y, down, "from up", false)
                    region.addStartPoint("from down", stairsDown.x, stairsDown.y)
                    return
                }
            }
        }
    }

    private fun randomRoom(): Room {
        val w = rp.roomMinWidth + random.nextInt(1 + rp.roomMaxWidth - rp.roomMinWidth)
        val h = rp.roomMinHeight + random.nextInt(1 + rp.roomMaxHeight - rp.roomMinHeight)
        val x = 1 + random.nextInt(region.width - w - 2)
        val y = 1 + random.nextInt(region.height - h - 2)
        return Room(region, x, y, w, h)
    }

    private class Room(val region: Region, val x: Int, val y: Int, val w: Int, val h: Int) {
        fun contains(cell: Cell) = cell.x >= x && cell.x < x + w && cell.y >= y && cell.y < y + h

        fun getRandomCell(random: Random): Cell {
            val xx = x + 1 + random.nextInt(w - 2)
            val yy = y + 1 + random.nextInt(h - 2)
            return region.getCell(xx, yy)
        }

        fun getMiddleCell(): Cell = region.getCell(x + w / 2, y + h / 2)

        fun addToRegion() {
            for (yy in 1..h - 1 - 1) {
                for (xx in 1..w - 1 - 1) {
                    region.getCell(x + xx, y + yy).setType(CellType.ROOM_FLOOR)
                }
            }
            for (xx in 0..w - 1) {
                region.getCell(x + xx, y).setType(CellType.ROOM_WALL)
                region.getCell(x + xx, y + h - 1).setType(CellType.ROOM_WALL)
            }
            for (yy in 0..h - 1) {
                region.getCell(x, y + yy).setType(CellType.ROOM_WALL)
                region.getCell(x + w - 1, y + yy).setType(CellType.ROOM_WALL)
            }
        }
    }

    class object : RegionGenerator {
        override fun generate(world: World, name: String, level: Int, up: String?, down: String?): Region =
            RoomFirstRegionGenerator(world, name, level, regionParameters(level), up, down).generate()

        private fun regionParameters(level: Int): RegionParameters {
            val rp = RegionParameters()
            if (level < 5) {
                rp.width = 80
                rp.height = 25
                rp.minRooms = 4
                rp.maxRooms = 10
                rp.roomMinWidth = 6
                rp.roomMaxWidth = 18
                rp.roomMinHeight = 5
                rp.roomMaxHeight = 9
            } else if (level < 10) {
                rp.width = 120
                rp.height = 30
                rp.minRooms = 6
                rp.maxRooms = 12
                rp.roomMinWidth = 6
                rp.roomMaxWidth = 18
                rp.roomMinHeight = 5
                rp.roomMaxHeight = 9
            } else if (level < 20) {
                rp.width = 160
                rp.height = 40
                rp.minRooms = 8
                rp.maxRooms = 16
                rp.roomMinWidth = 6
                rp.roomMaxWidth = 18
                rp.roomMinHeight = 5
                rp.roomMaxHeight = 9
            } else {
                rp.width = 200
                rp.height = 50
                rp.minRooms = 10
                rp.maxRooms = 25
                rp.roomMinWidth = 6
                rp.roomMaxWidth = 18
                rp.roomMinHeight = 5
                rp.roomMaxHeight = 9
            }
            return rp
        }
    }

    class RegionParameters {
        var width = 0
        var height = 0
        var minRooms = 4
        var maxRooms = 10
        var roomMinWidth = 6
        var roomMaxWidth = 18
        var roomMinHeight = 5
        var roomMaxHeight = 9
    }
}
